### ISSUE TITLE
win_lineinfile idempotence broken

### ISSUE THREAD

**Issue by one1zero1one** [ansible/ansible-modules-core#3381]
##### ISSUE TYPE

- Bug Report
##### COMPONENT NAME

- win_lineinfile
##### ANSIBLE VERSION

```
ansible 2.1.0 (devel cb7b3b489d) last updated 2016/03/30 15:14:21 (GMT +200)
  lib/ansible/modules/core: (detached HEAD 0268864211) last updated 2016/03/30 15:14:39 (GMT +200)
  lib/ansible/modules/extras: (detached HEAD 6978984244) last updated 2016/03/30 15:14:39 (GMT +200)
```
##### OS / ENVIRONMENT

ubuntu 14 -> windows 2012R2
##### SUMMARY

backrefs option doesn't arrive well inside the module, causing it not to function properly and also breaking idempotence (line will be always inserted at the end of the file on every execution, regardless of backrefs=yes and regardless of regexp match)
##### STEPS TO REPRODUCE

c:\test.txt containing 

```
test1
test2
test3
```

part of playbook

```
    win_lineinfile:
        dest: c:\test.txt
        regexp: "test2"
        line: "this will be added over and over"
        backrefs: yes
```
##### EXPECTED RESULTS

according to [this ](https://github.com/ansible/ansible/issues/4531)and similar to the behaviour on linux
i was expecting to have idempotent behaviour (to change the regexp into line if regexp found and do nothing if not found), which works on linux.

upon first execution, changed=1 
upon second execution, changed=0

```
test1
this will be added over and over
test3
```
##### ACTUAL RESULTS

upon first execution, changed=1 
upon second execution, changed=1

```
test1
this will be added over and over
test3
this will be added over and over
```
##### WORKAROUND/HACK

It seems that the problem comes from the backrefs variable arriving as True/False in the module, so this never gets triggered.

```
ElseIf ($backrefs -ne "no") {
                # No matches - no-op
```

and it forces it to go two elseifs down and add the line

I used this bit of code at the beginning of the module - at params  -  since I desperately needed the idempotent behavior :)

```
#dirty hack
$backrefs = Get-Attr $params "backrefs" "no";
if ( $backrefs -eq "True" ) {
$backrefs = "yes"
} else {
$backrefs = "no"
}
```

**Comment by jhawkesworth** [comment]
Modules should support yes/no or true/false for boolean parameters to align with the rest of Ansible.

Probably best to convert $backrefs into a boolean during parameter handling so tests can avoid fragile string comparison - ($backrefs -ne "no") - for example.

**Comment by ansibot** [comment]
@brianlloyd, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/jctanner/ansibullbot/blob/ISSUE_TRIAGE_CLASS/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@brianlloyd, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@brianlloyd, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@brianlloyd, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
@brianlloyd, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by jdevora** [comment]
And it doesn't do what it says in the documentation
http://docs.ansible.com/ansible/lineinfile_module.html
backrefs:  "... and if the regexp doesn't match anywhere in the file, the file will be left unchanged..."

**Comment by ansibot** [comment]
@brianlloyd, ping. This issue is still waiting on your response.
[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
This repository has been locked. All new issues and pull requests should be filed in https://github.com/ansible/ansible

Please read through the <a href="http://docs.ansible.com/ansible/dev_guide/repomerge.html#move-issues-and-prs-to-new-repo">repomerge</a> page in the dev guide. The guide contains links to tools which automatically move your issue or pull request to the ansible/ansible repo.

**Comment by dagwieers** [comment]
We have improved the win_lineinfile module a lot recently. Would it be possible to test against Ansible v2.3 or the devel branch, and report back ?

Thanks in advance !

**Comment by one1zero1one** [comment]
I can confirm the issue is not happening anymore, tested ansible 2.3.0.0 
Regards,

**Comment by dagwieers** [comment]
Thanks !

**Comment by rleigh-codelibre** [comment]
I'm still seeing the lack of idempotency:

```
- name: ci-base-setup (windows) | Export security policy
  win_command: "secedit /export /cfg {{ ansible_user_dir }}\\secpol.cfg"

- name: ci-base-setup (windows) | Update security policy to relax password rules and match Unix hosts
  win_lineinfile:
    dest: "{{ ansible_user_dir }}\\secpol.cfg"
    regexp: '^PasswordComplexity = 1$'
    line: 'PasswordComplexity = 0'
  register: secpol

- name: ci-base-setup (windows) | Import security policy
  win_command: "secedit /configure /db c:\\windows\\security\\local.sdb /cfg {{ ansible_user_dir }}\\secpol.cfg /areas SECURITYPOLICY"
  when: secpol.changed

- name: ci-base-setup (windows) | Delete exported security policy
  win_file:
    path: "{{ ansible_user_dir }}\\secpol.cfg"
    state: absent
```

`win_lineinfile` always reports that a change occurred.  Trying a few different regular expressions:

```
    regexp: '^PasswordComplexity.*=.*'
    regexp: '^PasswordComplexity =.*'
```
is OK -- no change is recorded as having been made

but

```
    regexp: '^PasswordComplexity = 1'
    regexp: '^PasswordComplexity = 1$'
```

always report a change.  Maybe this is just an expectation and/or documentation issue.  The working regexps match both the original and changed line, while the latter non-idempotent variants match only the original line.  Both cases are functionally identical in that the same input results in the same output, but is erroneously reported as being changed only for the less strict matches.

**Comment by jhawkesworth** [comment]
@rleigh-codelibre - please can you raise a new issue in ansible/ansible project if this problem still exists for you in ansible 2.3?  all the module code is now in the ansible/ansible github project and closed issues such as this are likely to be missed.

Many thanks,

Jon

**Comment by dagwieers** [comment]
@rleigh-codelibre That surely sounds like a bug. If that's with v2.3 we need you to open a new issue as @jhawkesworth suggested. We'll try to reproduce your example and prepare a fix to backport into v2.3.

Please refer your new issue here as well.

**Comment by rleigh-codelibre** [comment]
Will do.  This is with 2.3.0.

**Comment by rleigh-codelibre** [comment]
I've opened https://github.com/ansible/ansible/issues/25077 thanks.
